### ISSUE TITLE
How to really make startup code run before everything else?

### ISSUE THREAD

**Issue by gcmartinelli** [qtile/qtile#1392]
I have autostart setup as recommended in the documentation:
```# Autostart hook
@hook.subscribe.startup_once
def autostart():
    home = os.path.expanduser('~/.config/qtile/autostart.sh')
    subprocess.call([home])
```
My `autostart.sh` runs, among other things, `mons -e left` and `mons --primary DP-1`. DP-1 is my external monitor. This sets my external monitor as my main monitor.

The issue I'm having is that apparently, since startup is running on a subprocess, it finishes AFTER QTile has loaded my screens/widgets/etc... the effect is that my main screen is not set to my external monitor (as per my `autostart.sh` config). To get the effect I need to restart QTile after it has finished it's startup... then everything works.

I've tried running `os.system(home)` instead of `subprocess.call([home])` in the hopes of forcing QTile to run startup on the same process, but that had no effect.

I also tried putting all my config code inside a function with the `@hook.subscribe.startup_complete` decorator but that just made it so QTile wouldn't start.

~~My only solution right now is to force a `time.sleep()` after the startup decorator... but this is definitely not the best solution.~~ Solved my unique issue by running the specific code I needed inside my `config.py` file, although this (IMO) defeats part of the purpose of a autostart routine. Still curious to learn if there is a 'proper' way to work with the startup decorator.

TL;DR: I'm looking for a way to have Qtile run some code 'at startup' and only really start the rest of the config after that startup code is done.

**Comment by pawamoy** [comment]
I experienced the same issue. I eventually moved the code responsible for monitors settings into `~/.xinitrc` or `~/.xsession` (both are sh scripts, launched by `/usr/bin/startx` when the session opens). So in the end it has nothing to do with QTile!

As an example, here is my `~/.xsession`:
```sh
#!/bin/sh

# keyboard speed and latency
xset r rate 200 60

# setup my dual-screen
sh ~/.screenlayout/home.sh

# window effects (transparency, blur, shadows, etc.)
compton --config ~/.config/compton/compton.conf &

# start qtile
exec qtile
```

This way you effectively setup your monitors before QTile launch, so you don't have to reload it a second time to fix the screens.

**Comment by m-col** [comment]
A `main` function defined in your config is executed ([here](https://github.com/qtile/qtile/blob/master/libqtile/core/manager.py#L100)) before Qtile really does anything, serving as a `startup_before` equivalent. The only real step before that would be inside the `xinitrc` and `xsession` files as you mention - which are typically used for things such as monitor configuration.

**Comment by tych0** [comment]
Gonna go ahead and close this since it seems answered.
